pennylane/compiler/python_compiler/transforms/decompose.py

```diff
@@ -0,0 +1,285 @@
+
+    def gate_set_contains(self, op: Operator) -> bool:
+        """Check if the operator is in the gate set."""
+        return op.name in self.gate_set
+
+    def stopping_condition(self, op: Operator) -> bool:
+        """Function to determine whether an operator needs to be decomposed or not."""
+        if not op.has_decomposition:
+            if not self.gate_set_contains(op):
+                warnings.warn(
+                    f"Operator {op.name} does not define a decomposition and was not "
+                    f"found in the target gate set. To remove this warning, add the operator "
+                    f"name ({op.name}) or type ({type(op)}) to the gate set.",
+                    UserWarning,
+                )
+            return True
+        return self.gate_set_contains(op)
+
+    def decompose_operation(self, op: Operator):
+        """Decompose the operation if it is not in the gate set."""
+        if self.gate_set_contains(op):
+            return [op]
+        decomposition = list(
+            _operator_decomposition_gen(
+                op,
+                self.stopping_condition,
+                max_expansion=self.max_expansion,
+                current_depth=0,
+                decomp_graph=None,
+            )
+        )
+        return decomposition
+
+    def resolve_constant_params(self, param: SSAValue) -> Union[float, int]:
+        """Resolve the constant parameter from the SSA value."""
+        op = param.owner
+        if not isinstance(op, ConstantOp):
+            raise NotImplementedError(f"Expected ConstantOp but got {type(op)}")
+        val = op.value
+        if not isinstance(val, (FloatAttr, IntegerAttr)):
+            raise NotImplementedError(f"Constant has unexpected attr type: {type(val)}")
+        if val not in self.params_to_ssa_params:
+            self.params_to_ssa_params[val] = param
+        return val.value.data
+
+    def ssa_to_qml_params(self, op: CustomOp) -> list[float | int]:
+        """Get the parameters from the operation."""
+        if not hasattr(op, "params"):
+            return []
+        return [self.resolve_constant_params(p) for p in op.params if p is not None]
+
+    def ssa_to_qml_wires(self, op: CustomOp) -> list[int]:
+        """Get the wires from the operation."""
+        if not hasattr(op, "in_qubits"):
+            return []
+        return [self.resolve_constant_wire(q) for q in op.in_qubits if q is not None]
+
+    def resolve_constant_wire(self, in_qubit: SSAValue) -> int:
+        """Resolve the wire for the given SSA qubit."""
+        op = in_qubit.owner
+        if isinstance(op, ExtractOp):
+            return op.idx_attr.parameters[0].data
+        if isinstance(op, ConstantOp):
+            val = op.value
+            return val.value.data
+        if isinstance(op, CustomOp):
+            # The CustomOp should be the last one that updated
+            # the mapping for this SSA qubit.
+            wire_position = in_qubit.index
```

Comment:
This will not work how you expect. `in_qubit` here is an `OpResult` instance, and the `index` field of this class refers to the index of the `OpResult` in the `results` of the owner of the `OpResult` ([ref](https://github.com/xdslproject/xdsl/blob/8fa44a3cc28d0fcfcabc89c3c412d29df32aa91e/xdsl/ir/core.py#L652)). For qubits that are returned by single qubit `CustomOp`s, this will always be 0 regardless of which wire the op is applied to, since each single qubit `CustomOp` gives a single qubit as its result in `out_qubits`.